File: kotlin-typescript/karakum/plugins/convertConflictingOverloads.js
import ts from "typescript";

function isConflictingOverload(node) {
    return ts.isIdentifier(node.name) && (
        (
            node.name.text === "applyCodeActionCommand"

            && (
                (
                    node.parameters[0]
                    && ts.isUnionTypeNode(node.parameters[0].type)
                )
                || (
                    node.parameters[1]
                    && ts.isUnionTypeNode(node.parameters[1].type)
                )
            )
        )
    )
}

export default function (node, context, render) {
    if (
        ts.isMethodSignature(node)
        && isConflictingOverload(node)
    ) {
        return ""
    }

    if (
        ts.isCallSignatureDeclaration(node)
        && node.parent
        && ts.isInterfaceDeclaration(node.parent)
        && node.parent.name.text === "NodesVisitor"

        && node.typeParameters
        && node.typeParameters.length === 3
    ) {
        return ""
    }

    if (
        ts.isFunctionDeclaration(node)
        && node.name
        && node.name.text === "visitNodes"

        && node.typeParameters
        && node.typeParameters.length === 3
    ) {
        return ""
    }

    return null
}

File: kotlin-typescript/karakum/src/jsMain/kotlin/typescript/karakum/plugins/convertConflictingOverloads.kt
package typescript.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.util.getParentOrNull
import typescript.MethodSignature
import typescript.asArray
import typescript.isCallSignatureDeclaration
import typescript.isFunctionDeclaration
import typescript.isIdentifier
import typescript.isInterfaceDeclaration
import typescript.isMethodSignature
import typescript.isUnionTypeNode

private fun isConflictingOverload(node: MethodSignature) = nullable {
    val name = node.name
    ensure(isIdentifier(name))
    ensure(name.text == "applyCodeActionCommand")

    ensureNotNull(
        nullable {
            val firstParameter = ensureNotNull(node.parameters.asArray().firstOrNull())

            val firstParameterType = ensureNotNull(firstParameter.type)
            ensure(isUnionTypeNode(firstParameterType))
        } ?: nullable {
            val secondParameter = ensureNotNull(node.parameters.asArray().getOrNull(1))

            val secondParameterType = ensureNotNull(secondParameter.type)
            ensure(isUnionTypeNode(secondParameterType))
        }
    )
} != null

val convertConflictingOverloads = createPlugin { node, _, _ ->
    nullable {
        ensure(isMethodSignature(node))
        ensure(isConflictingOverload(node))

        ""
    } ?: nullable {
        ensure(isCallSignatureDeclaration(node))

        val interfaceNode = ensureNotNull(node.getParentOrNull())
        ensure(isInterfaceDeclaration(interfaceNode))
        ensure(interfaceNode.name.text == "NodesVisitor")

        ensure(node.typeParameters?.asArray()?.size == 3)

        ""
    } ?: nullable {
        ensure(isFunctionDeclaration(node))
        ensure(node.name?.text == "visitNodes")

        ensure(node.typeParameters?.asArray()?.size == 3)

        ""
    }
}
